sdk/core/core/src/az_client_secret_credential.c

```diff
@@ -43,125 +43,75 @@ static AZ_NODISCARD az_result az_token_credential_get_token(
-
-  AZ_CONTRACT_ARG_VALID_SPAN(credential->client_id);
-  AZ_CONTRACT_ARG_VALID_SPAN(credential->client_secret);
-
   {
-    AZ_CONTRACT(resource_url.size >= 12, AZ_ERROR_ARG);
-    AZ_CONTRACT(credential->tenant_id.size >= 32, AZ_ERROR_ARG);
-    AZ_CONTRACT(credential->client_id.size >= 32, AZ_ERROR_ARG);
-    AZ_CONTRACT(credential->client_secret.size >= 32, AZ_ERROR_ARG);
-  }
-
-  static az_span const auth_url1 = AZ_CONST_STR("https://login.microsoftonline.com/");
-  static az_span const auth_url2 = AZ_CONST_STR("/oauth2/token");
-
-  static az_span const auth_body1 = AZ_CONST_STR("grant_type=client_credentials&client_id=");
-  static az_span const auth_body2 = AZ_CONST_STR("&client_secret=");
-  static az_span const auth_body3 = AZ_CONST_STR("&resource=");
+    uint8_t auth_url_buf[100] = { 0 };
+    az_span auth_url = { 0 };
+    {
+      static az_span const auth_url1 = AZ_CONST_STR("https://login.microsoftonline.com/");
+      static az_span const auth_url2 = AZ_CONST_STR("/oauth2/token");
 
-  size_t const auth_url_maxsize = auth_url1.size
-      + (credential->tenant_id.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR)
-      + auth_url2.size;
+      az_span_builder builder
+          = az_span_builder_create((az_mut_span)AZ_SPAN_FROM_ARRAY(auth_url_buf));
 
-  AZ_CONTRACT(auth_url_maxsize <= (size_t) ~(uint16_t)0, AZ_ERROR_ARG);
+      AZ_RETURN_IF_FAILED(az_span_builder_append(&builder, auth_url1));
+      AZ_RETURN_IF_FAILED(az_uri_encode(credential->tenant_id, &builder));
+      AZ_RETURN_IF_FAILED(az_span_builder_append(&builder, auth_url2));
 
-  size_t const headers_size = sizeof(az_pair) + 7; // We need 7 because our code aligns at 8 byte boundary
-  {
-    AZ_CONTRACT(
-        response_buf.size >= AZ_TOKEN_CREDENTIAL_GET_TOKEN_MIN_BUFFER, AZ_ERROR_BUFFER_OVERFLOW);
-
-    size_t const request_elements[] = {
-      credential->tenant_id.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-      credential->client_id.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-      credential->client_secret.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-      resource_url.size * AZ_TOKEN_CREDENTIAL_GET_TOKEN_URLENCODE_FACTOR,
-    };
-
-    size_t required_request_size = auth_url1.size + auth_url2.size + auth_body1.size
-        + auth_body2.size + auth_body3.size + headers_size;
-
-    for (size_t i = 0; i < AZ_ARRAY_SIZE(request_elements); ++i) {
-      required_request_size += request_elements[i];
-      AZ_CONTRACT(required_request_size > request_elements[i], AZ_ERROR_BUFFER_OVERFLOW);
+      auth_url = az_span_builder_result(&builder);
     }
 
-    AZ_CONTRACT(response_buf.size >= required_request_size, AZ_ERROR_BUFFER_OVERFLOW);
-  }
+    uint8_t auth_body_buf[200] = { 0 };
+    az_span auth_body = { 0 };
 
-  {
-    AZ_CONTRACT(!az_span_is_overlap(az_mut_span_to_span(response_buf), resource_url), AZ_ERROR_ARG);
+    static az_span const auth_body1 = AZ_CONST_STR("grant_type=client_credentials&client_id=");
```

Comment:
auth_body_granttype
auth_body_clientsecret
auth_body_resource
etc...

Reply:
Done - the code is rewritten in a way it is no longer needed.